### DISCUSSION TITLE
[v5] How can i import a service from the IoC inside provider boot() method ?

### DISCUSSION THREAD

**Discussion by wemersonrv** [adonisjs/core#3095] [Help] [answered]
Hello guys, greetings from Brazil.

I'm recreating an Adonis 4 project from scratch with Adonis 5, and need to port an IoC implementation. The service and the provider works fine... Just need to call it from the provider `boot` method to start it BEFORE any HTTP request.

## My v4 Service Provider
Here my provider on Adonis v4... it works fine when i call it from boot().

```js
'use strict'

const path = require('path')
const Helpers = use('Helpers')
const { ServiceProvider, ioc } = require('@adonisjs/fold')

class MyProvider extends ServiceProvider {
  register() {
    // Do not run it on ACE commands
    if (Helpers.isAceCommand()) return

    ioc.singleton('MyService', function (app) {
      const myPath = path.join(__dirname, '../app/Services/MyService')
      
      const mySingleton = require(myPath)

      return new mySingleton()
    })
  }

  boot() { 
    // Do not run it on ACE commands
    if (Helpers.isAceCommand()) return

    const initMyService = use('MyService')
  }
}

module.exports = MyProvider
```

## Trying to make the same in new project with Adonis V5
```js
import { ApplicationContract } from '@ioc:Adonis/Core/Application'

import MyService from 'App/Services/MyService'

export default class MyProvider {
  constructor(protected app: ApplicationContract) {}

  public register() {
    this.app.container.singleton('MyService', function () {
      return new myService()
    })
  }

  public async boot() {
    //// ( 1 )  don't know how to call the service here as i do on v4 version, because cannot be imported
  }

  public async ready() {
    
  }

  public async shutdown() {
    
  }
}
``` 

Tested here and this code works, but only when the `@ioc:MyService` is imported at first time. So i just need to understand how to import it inside the provider's `boot` method. There is a way to do that?

**Comment by sergefabre** [comment]
If that can help you : https://github.com/adonisjs/core/discussions/3101

**Reply by wemersonrv** [reply]
> If that can help you : #3101

This link on the thread, is the on e i used to understand the v5 flow for IoC... it works, but not 100% as i need. 

1. Add the service to the IoC with a Service Provider: **DONE**
2. import it or use **INJECT** on my controller as is show on this link: **DONE**
3. Need to create the instance on provider boot time: **NOT IMPLEMENTED** i need this.

It works... BUT will only create the service instance (singleton) after receive the first `http request` that calls the controller/method... and i need to make this on service provider `boot` time **BEFORE** receive any request. See my v4 service provider that i post when create this thread, it works on v4.

**Comment by wemersonrv** [comment, marked as answer]
Found it... **`this.app.container.use('@ioc:App/Services/MyService')`**

Below my provider complete:

```js
import { ApplicationContract } from '@ioc:Adonis/Core/Application'
import MyService from 'App/Services/MyService'

export default class MyProvider {
  constructor(protected app: ApplicationContract) {}

  public register() {
    this.app.container.singleton('MyService', function () {
      return new myService()
    })
  }

  public async boot() {
    ////// SOLUTION
    this.app.container.use('@ioc:App/Services/MyService')
  }

  public async ready() {
    
  }

  public async shutdown() {
    
  }
}
```

**Reply by sergefabre** [reply]
new in https://github.com/adonisjs/core/discussions/3101

**Reply by wemersonrv** [reply]
> new in #3101

I use `@inject` on my controllers, on entire class and on methods and works fine. But not works on provider `boot` method... so i use `this.app.container.use('@ioc:App/Services/MyService')` instead..

**Reply by sergefabre** [reply]
Hum...
my code not work. 
It doesn't make a Singleton. The constructor is call everytime... :(

**Reply by wemersonrv** [reply]
> Hum...
> my code not work.
> It doesn't make a Singleton. The constructor is call everytime... :(


Well, it has to work if is singleton or not... Can you share your controller code ?

Here i had some issues when injecting in a method, must add the `ctx` argument, even if don't use it...

```ts
@inject(['@ioc:App/Services/WaaaService'])
export default class WaaaController {
  constructor(protected _entireController: WaaaServiceContract) {
    console.log(this_entireController.startedAt) // 31/08/2021 10:23:32  (Brazilian date format) 
  }

  @inject([null, '@ioc:App/Services/WaaaService']) // ctx cames first as null
  public index({}, methodOnly: WaaaServiceContract) {
    console.log(methodOnly.startedAt) // 31/08/2021 10:23:32  (Brazilian date format) 
  }
}
```

Here as i don't use ctx on this method, i need to add at least as empty object be
